### PR TITLE
add spr! to LinearAlgebra.BLAS

### PR DESCRIPTION
**Pull request by bjarthur**
i used `spmv!` as a template

### CODE REVIEW

File: stdlib/LinearAlgebra/src/blas.jl [hunk 1]
```diff
@@ -1145,6 +1146,71 @@ Return the updated `y`.
...
+
+function spr!(uplo::AbstractChar,
+              α::Real, x::Union{DenseArray{T}, AbstractVector{T}},
+              AP::Union{DenseArray{T}, AbstractVector{T}}) where {T <: BlasReal}
+    require_one_based_indexing(AP, x)
+    N = length(x)
+    if 2*length(AP) < N*(N + 1)
+        throw(DimensionMismatch("Packed symmetric matrix A has size smaller than length(x) = $(N)."))
+    end
+    return spr!(uplo, N, convert(T, α), x, stride(x, 1), AP)
```
Comment (sostock) [inline stdlib/LinearAlgebra/src/blas.jl:1188]: This will use the wrong pointer for `x` if `stride(x,1)` is negative. However, no BLAS wrappers except for `gemv!` currently handle negative strides correctly, so it may be ok to leave this for a separate PR.

(Support for negative strides in `gemv!` was added in #41513 and I plan to add it for the other BLAS functions once there is a decision on whether to merge #42012 or #42054.)

File: stdlib/LinearAlgebra/src/blas.jl [hunk 2]
```diff
@@ -1145,6 +1146,71 @@ Return the updated `y`.
...
+                  AP,
+                  1)
+            return AP
+        end
+    end
+end
+
+function spr!(uplo::AbstractChar,
+              α::Real, x::Union{DenseArray{T}, AbstractVector{T}},
+              AP::Union{DenseArray{T}, AbstractVector{T}}) where {T <: BlasReal}
```
[the remark this answers is not in the document]
Comment (bjarthur) [reply stdlib/LinearAlgebra/src/blas.jl:1182]: i too wondered about `x::DenseArray`.  but that is what `spmv!` inputs.  is there a use case for it there?  `smpv!` also does not check the stride of AP.  happy to change it here, but would be nice to keep everything in sync.

**Comment by dkarrasch** [comment]
I think this PR is complete and consistent with the current state of the surrounding functions, so I'll merge once tests pass. I believe we have an issue tracking negative strides, another issue raised was to check the stride for the vector inputs and to catch non-sense `uplo` arguments. These should be ideally dealt with in a consistent fashion in other PRs.

**Comment by dkarrasch** [comment]
> @iolaszlo in your [PR for spmv](https://github.com/JuliaLang/julia/pull/34320) you included `DenseArray` in the method signature. what is the use case for that? for my PR for `spr` i used your code as a template, and folks are wondering why not just use `AbstractVector` alone. thx.

Some discussion can be found here: https://github.com/JuliaLang/julia/pull/34320#discussion_r364594010. I think there are several places in the BLAS context where we consider any contigious array simply as a list, no matter what its dimensionality is "officially".

File: stdlib/LinearAlgebra/src/blas.jl [hunk 3]
```diff
@@ -1145,6 +1146,71 @@ Return the updated `y`.
...
+    require_one_based_indexing(AP, x)
+    N = length(x)
```
[the remark this answers is not in the document]
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/blas.jl:1184]: I started that kind of blas-global clean-up in a separate PR, so let's consider this as resolved.

**Comment by fkastner** [comment]
The docstring seems wrong. It says 

> Update matrix A as α*A*x*x',[...]

while it should be `A + α*x*x'`, iiuc.
